Refactor method typing

fo5for opened the pull request:
The code for method typing is refactored to remove the use of mutable maps for storing method signatures and minimize messy substitutions of signatures. Below is an overview of the major changes:

**Refactor TypeDef and TypeRef**:
- `TypeDef` now stores the typed type of its body instead of the syntactic type.
- In order to type the body of a type definition before the `TypeDef` is added to the `Ctx`, `typeType` now takes an additional optional parameter `newDefsInfo` with the kind and number of type paramerters of the new definitions.
- The auxilliary `typeType2` also returns the list of `TypeVariables` used to instantiate implicit `TypeVar`s in the body, which are instantiated by substitution when a `TypeRef` is expanded.
- `TypeRef` now takes a `TypeName` as the base instead of a `TypeDef`, and no longer stores a `Ctx`, instead requiring a `Ctx` argument to retrieve the definition for expansion.
- `TypeRef.expand` no longer takes a `Raise` argument as it only performs substitution instead of typing the body every time. This eliminated much of the error swallowing.
- Caveat: `SimpleType.<:<` also needs to take a `Ctx` argument, which is problematic for normal forms as they implement `<:<` with `toType` and `SimpleType.<:<`, but `toType` would never produce a `TypeRef`. An "empty" `Ctx` is pass to `SimpleType.<:<` for now.

**Refactor method typing**:
- Removed mutable maps in context used for storing method signatures.
- Moved method bindings to a separate environment. Beside the actual binding, the inferred type is also stored with key delimited by `#`. This is only used for printing in `DiffTests`.
- Instead of looking up the signatures of inherited methods and substituting type variables for the type arguments, `typeMethod` now recursively traverse the type definition body, and collect inherited method signatures by typing them with the proper arguments applied.

New subclasses of `PolymorphicType`:
- `MethodType`: stores the parents of the method in addition. `&` intersects the bodies of two inherited methods and concatenate their parents; `+` overwrites a binding with the new signature with the same parent, or concatenate the parents if the call is ambiguous.
- `AbstractConstructor`: stores the set of abstract methods for printing when a constructor call of an abstract type is attempted. This is needed since it is no longer possible to look up the list of all declared and defined methods after the `TypeDef` is processed from the mutable maps.

**Miscellaneous improvements**:
- `checkRegular` no longer expands `TypeRef`s with parameter tags, which eliminates duplicate errors.
- A polymorphism issue for methods was accidentally identified and fixed by the refactoring in `MethodAndMatches.mls`

fo5for committed:
Refactor `TypeDef` and `TypeRef`

- `TypeDef` now stores the typed type of its body instead of the
  syntactic type.
- In order to type the body of a type definition before the `TypeDef`
  is added to the `Ctx`, `typeType` now takes an additional optional
  parameter `newDefsInfo` with the kind and number of type paramerters
  of the new definitions.
- The auxilliary `typeType2` also returns the list of `TypeVariables`
  used to instantiate implicit `TypeVar`s in the body, which are
  instantiated by substitution when a `TypeRef` is expanded.
- `TypeRef` now takes a `TypeName` as the base instead of a `TypeDef`,
  and no longer stores a `Ctx`, instead requiring a `Ctx` argument to
  retrieve the definition for expansion.
- `TypeRef.expand` no longer takes a `Raise` argument as it only
  performs substitution instead of typing the body every time. This
  eliminated much of the error swallowing.
- Caveat: `SimpleType.<:<` also needs to take a `Ctx` argument, which
  is problematic for normal forms as they implement `<:<` with
  `toType` and `SimpleType.<:<`, but `toType` would never produce a
  `TypeRef`. An "empty" `Ctx` is pass to `SimpleType.<:<` for now.

Miscellaneous improvements:
- `checkRegular` no longer expands `TypeRef`s with parameter tags,
  which eliminates duplicate errors.
- A polymorphism issue for methods was accidentally identified and
  fixed by the refactoring in `MethodAndMatches.mls`

fo5for committed:
Refactor method typing

- Removed mutable maps in context used for storing method signatures.
- Moved method bindings to a separate environment. Beside the actual binding, the inferred type is also stored with key delimited by `#`. This is only used for printing in `DiffTests`.
- Instead of looking up the signatures of inherited methods and substituting type variables for the type arguments, `typeMethod` now recursively traverse the type definition body, and collect inherited method signatures by typing them with the proper arguments applied.

New subclasses of `PolymorphicType`:
- `MethodType`: stores the parents of the method in addition. `&` intersects the bodies of two inherited methods under the new parent; `+` overwrites a binding with the new signature with the same parent, or concatenate the parents if the call is ambiguous.
- `AbstractConstructor`: stores the set of abstract methods for printing when a constructor call of an abstract type is attempted. This is needed since it is no longer possible to look up the list of all declared and defined methods after the `TypeDef` is processed from the mutable maps.

fo5for committed:
Fix multiple inherited method definitions

fo5for committed:
Fix parent check in subsumption check

fo5for committed:
Cleanup

LPTK commented:
Awesome improvements! Can you please rebase your branch on top of the current `mlscript` main branch?
We can keep your commits as they seem fairly atomic. Each of them passes all the tests, right?

LPTK commented on shared/src/main/scala/mlscript/NormalForms.scala:
```diff
@@ -159,7 +160,7 @@ class NormalForms extends TyperDatatypes { self: Typer =>
       case _: RhsField | _: RhsBases => N
     }
-    def <:< (that: RhsNf): Bool = this.toType() <:< that.toType() // TODO less inefficient! (uncached calls to toType)
+    def <:< (that: RhsNf): Bool = this.toType().<:<(that.toType())(Ctx.empty) // TODO less inefficient! (uncached calls to toType)
```
```suggestion
    def <:< (that: RhsNf): Bool = (this.toType() <:< that.toType())(Ctx.empty) // TODO less inefficient! (uncached calls to toType)
```

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/NormalForms.scala:
```diff
@@ -202,7 +203,7 @@ class NormalForms extends TyperDatatypes { self: Typer =>
     def & (that: Conjunct): Opt[Conjunct] =
       // trace(s"?? $this & $that ${lnf & that.lnf} ${rnf | that.rnf}") {
-      if (lnf.toType() <:< that.rnf.toType()) N // TODO support <:< on any Nf? // TODO less inefficient! (uncached calls to toType)
+      if (lnf.toType().<:<(that.rnf.toType())(Ctx.empty)) N // TODO support <:< on any Nf? // TODO less inefficient! (uncached calls to toType)
```
```suggestion
      if (lnf.toType() <:< that.rnf.toType())(Ctx.empty)) N // TODO support <:< on any Nf? // TODO less inefficient! (uncached calls to toType)
```

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
   type Binding = Str -> TypeScheme
   type Bindings = Map[Str, TypeScheme]
-  
+
```
```suggestion
  
```

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
-    targs: List[TypeVariable],
-    body: Type,
+    tparamsargs: List[(TypeName, TypeVariable)],
+    tvars: Iterable[TypeVariable], // "implicit" type variables. instantiate every time a `TypeRef` is expanded
```
Why `Iterable`? I try to avoid using such abstract collections as they can obscure the underlying data structure shape and complexity of operations we do on them.

fo5for replied:
I used the return type of `Map.values` (in `typeType2`) to avoid creating unnecessary collections. This would be removed when we fix the soundness issue with type variables.

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
           ctx.tyDefs.get(v.name).fold(Set.empty[Var])(_.allBaseClasses(ctx)(traversed + v)))
-    val targsMap: Map[Str, TypeVariable] = tparams.map(_.name).zip(targs).toMap
+    val tparams: List[TypeName] = tparamsargs.map(_._1)
+    val targs: List[TypeVariable] = tparamsargs.map(_._2)
```
```suggestion
    val (tparams: List[TypeName], targs: List[TypeVariable]) = tparamsargs.unzip
```

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+    def wrapMethod(bodyTy: PolymorphicType)(implicit prov: TypeProvenance): MethodType =
+      MethodType(bodyTy.level, FunctionType(singleTup(thisTy), bodyTy.body)(prov), nme)
+  }
+
```
```suggestion
  
```

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
-    val targsMap: Map[Str, TypeVariable] = tparams.map(_.name).zip(targs).toMap
+    val tparams: List[TypeName] = tparamsargs.map(_._1)
+    val targs: List[TypeVariable] = tparamsargs.map(_._2)
+    def thisTy(implicit prov: TypeProvenance): TypeRef = TypeRef(nme, targs)(prov)
```
Seems to be only used in `wrapMethod`, so would probably be better as a local `val`.

fo5for replied:
The `TypeRef` carries the same prov as the `MethodType`, which is different for every method. Making this a `val` would require changing it to a `noProv`, which loses some information for error reporting.

LPTK replied:
I'm not suggesting to make it a `val` _field_. I meant a _local_ `val` in `wrapMethod`.

fo5for replied:
Oh... 😅 Addressed in c5f6b0817ec9905325a697a9281652782b2d2e30

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+        this.defns ++ that.defns
+      )
+    }
+    private def addprt(prt: TypeName): MethodDefs =
```
Method naming consistency:
```suggestion
    private def addParent(prt: TypeName): MethodDefs =
```

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+
+  private case class MethodDefs(
+    tn: TypeName,
+    prts: List[MethodDefs],
```
I'd rather call this `parents`. Took me a while to get what `prts` meant 😄

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+          case mn -> defns =>
+            implicit val prov: TypeProvenance = TypeProvenance(tn.toLoc, "inherited method declaration")
+            S(mn -> MethodType(defns.head.level,
+              err(msg"A method definition must be given when inheriting multiple method definitions" -> tn.toLoc
```
```suggestion
              err(msg"An overriding method definition must be given when inheriting from multiple method definitions" -> tn.toLoc
```

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -186,14 +220,17 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+    case S(td) =>
+      require(td.kind is Cls)
+      ClassTag(Var(td.nme.name.decapitalize), ctx.allBaseClassesOf(td.nme.name))(prov)
+    case N => errType(prov)
```
This change seems unnecessary and may silence genuine bugs.

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -250,6 +283,8 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
   def processTypeDefs(newDefs0: List[mlscript.TypeDef])(implicit ctx: Ctx, raise: Raise): Ctx = {
     var allDefs = ctx.tyDefs
     var allEnv = ctx.env.clone
+    var allMthEnv = ctx.mthenv.clone
```
```suggestion
    val allEnv = ctx.env.clone
    val allMthEnv = ctx.mthenv.clone
```

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,94 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+
   case class TypeDef(
     kind: TypeDefKind,
     nme: TypeName,
```
This should probably be a `Var` instead.

LPTK replied:
Actually `TypeName` is fine. This name serves both as a type name and a term name (for the constructor), but it's primarily a type one.

fo5for replied:
You probably meant the one in `MethodDef`, which is addressed in 0fcee02d805237ba9580d4eb0154da5b1bb0c69a

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +393,232 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+          val fullName = s"${tn.name}.${nme}"
+          println(s">> Checking subsumption for declared type of $nme : $mt")
+          declsInherit.get(nme).orElse(defnsInherit.get(nme).tap(_.foreach { mt2 =>
+            err(msg"Overriding method ${mt2.prts.headOption.fold("")(_.name)}.${nme} without explicit declaration is not allowed."
```
As discussed in the meeting, this `mt2.prts.headOption.fold("")(_.name)` is a bit fishy. Maybe better done as `mt2.prts.headOption.foreach { prt => ... }`.

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +393,232 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+          decls.get(nme).orElse((declsInherit.get(nme), defnsInherit.get(nme)) match {
+            case (S(decl), S(defn)) if defn.prts.toSet.subsetOf(decl.prts.toSet) => S(decl)
+            case (_, S(defn)) =>
+              err(msg"Overriding method ${defn.prts.headOption.fold("")(_.name)}.${nme} without explicit declaration is not allowed."
```
As discussed in the meeting, we could do some deduplication with the above here, at least for this specific error-reporting part.

fo5for replied:
Addressed in f48c96fcda9dfbf2f7633ccfec76d4387d9aa354

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +393,232 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+          val mthTy = td.wrapMethod(mt)
+          if (!ctx.containsmth(fullName) || !decls.isDefinedAt(nme)) ctx.mthenv += fullName -> mthTy
+          ctx.mthenv += s"${tn.name}#${nme}" -> mthTy
+          if (!decls.isDefinedAt(nme) && !declsInherit.isDefinedAt(nme)) registerImplicit(nme, mthTy)
```
It would be good to add some (very brief) documentation for this logic and in the two lines above. When I read it first I had trouble understanding it.

fo5for replied:
You were right to have trouble! Some of the conditions are redundant, which have been fixed in dbf9e69d1f0b3293e36d97c6cf378bdfdb422fa8. The `#` line should now be more readable after structured keys are used for `mthEnv`.

LPTK replied:
Cool, but I'd still like some doc!! The semantics of methods may be crystal clear to you, but imagine how much trouble someone who is not sure will have, reading this code.

(I'm not saying the code _I_ write isn't similarly hard to understand BTW – it also needs to be better documented.)

fo5for replied:
Actually, it was not crystal clear to me either. I realized that at some point during this refactoring, and I drew a graph for what to do for the 16 cases where a subset of inherited declaration, inherited definition, new declaration and new definition are present. 😅 Documenting them in text would be a huge mess but I'm not sure if including graphics in the source for documentation is a good idea 😅

LPTK replied:
Okay, I find it much better now 👍

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +393,232 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+        }
+        defns.foreach { case nme -> mt => 
+          implicit val prov: TypeProvenance = mt.prov
+          val fullName = s"${tn.name}.${nme}"
```
Let's use proper structured keys instead of raw strings here, as discussed in the meeting. With a doc comment on the meaning of the shapes of the keys.

fo5for replied:
Addressed in 78d1a5ef3b58faaaa85466519f26dad19f5524c0

LPTK commented on shared/src/main/scala/mlscript/TyperDatatypes.scala:
```diff
@@ -31,6 +31,36 @@ abstract class TyperDatatypes extends TyperHelpers { self: Typer =>
+    def unapply(mt: MethodType): S[(Int, SimpleType, List[TypeName])] = S((mt.level, mt.body, mt.prts))
+  }
+
+  class AbstractConstructor(val absMths: Set[TypeName])(body: SimpleType) extends PolymorphicType(0, body) {
```
Not a big fan of this subtype. Its semantics relies on downstream code to _remember_ to check for it, and if they forget, it will silently do the wrong thing. It's unclear why it needs to be a subtype of `PolymorphicType` at all. For instance, does the `TyperHelpers` method below make sense for it?

```scala
  def subst(ts: PolymorphicType, map: Map[SimpleType, SimpleType]): PolymorphicType = 
    PolymorphicType(ts.level, subst(ts.body, map))
```

Probably not.

Better make this and `TypeScheme` both subclasses of a new `TypeInfo` base class, and make typing environments map variable names to `TypeInfo`s, to be checked in the `Var` typing case later on.

fo5for replied:
Addressed in 822bcd4ac635eb10660981cdb4d39cb501d3d8cb

LPTK commented on shared/src/main/scala/mlscript/TyperDatatypes.scala:
```diff
@@ -31,6 +31,36 @@ abstract class TyperDatatypes extends TyperHelpers { self: Typer =>
     def rigidify(implicit lvl: Int): SimpleType = freshenAbove(level, body, rigidify = true)
   }
   
+  class MethodType(level: Int, body: SimpleType, val prts: List[TypeName], val single: Bool)
```
I'm also uneasy about this being a subclass of `PolymorphicType`. Typing environments never map to this, right? Maybe a different base trait/class can be used for the purposes of method type handling.

fo5for replied:
Addressed in d7e4fecb249181bed9b09864e3ca35b645bbd7d8

LPTK reviewed (changes requested):
It looks pretty good overall, except for the few comments I added.

Another thing: I think the following should be disallowed:
```scala
class C
    method F: int
    method F = 1
//│ Defined class C
//│ Declared C.F: C -> int
//│ Defined C.F: C -> 1

class D: C
    method F: 1
//│ Defined class D
//│ Declared D.F: D -> 1
```
Indeed, the whole point of requiring a signature for overridden definitions is that the original base class may later change the method _definition_ (leaving the declaration unchanged) without breaking downstream code. But in this example, if class `C` changes `F` to return `2`, it will actually break downstream code!

fo5for committed:
Cleanup and renameing for consistency

fo5for committed:
Change method identifiers to `Var`s

fo5for committed:
Use structured keys for `mthEnv`

fo5for committed:
Fix type hierarchy of `AbstractConstructor`

fo5for committed:
Fix type hierarchy of `MethodType`

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -36,62 +36,72 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
-    def ++=(bs: IterableOnce[Binding]): Unit = bs.iterator.foreach(+=)
-    def get(name: Str): Opt[TypeScheme] = env.get(name) orElse parent.dlof(_.get(name))(N)
-    def getmth(name: Str): Opt[MethodType] = mthenv.get(name) orElse parent.dlof(_.getmth(name))(N)
+    def +=(b: Str -> TypeInfo): Unit = env += b
```
Why expand the `Binding` alias everywhere?

fo5for replied:
This was needed when the `TypeInfo` base class was introduced, and `env` was changed to map to `TypeInfo`s. We could either expand it here, or change the definition of the `Binding` alias and expand the reference in the return type of `typeStatement`, which only returns and is required to return a binding to a `TypeScheme`. I chose the former based on my interpretation of the word "binding", as a `TypeInfo` is not strictly a "binding". I can change it to the other way if you think that's more appropriate.

LPTK replied:
Hmm okay, I guess that's fine.

LPTK commented on shared/src/main/scala/mlscript/TyperDatatypes.scala:
```diff
@@ -31,34 +35,30 @@ abstract class TyperDatatypes extends TyperHelpers { self: Typer =>
   
-  class MethodType(level: Int, body: SimpleType, val prts: List[TypeName], val single: Bool)
-      extends PolymorphicType(level, body) {
+  class MethodType(val level: Int, val body: Opt[SimpleType], val parents: List[TypeName], val single: Bool)
```
You could make this `case class MethodType(level: Int, body: Opt[SimpleType], parents: List[TypeName], single: Bool)` and avoid having to implement `copy`, `apply`, and `unapply`. Also, could you document what `single` is for?

fo5for replied:
I wrote it this way to avoid explicitly dealing with `single` outside the class, which can only be turned into `false` by `&`. The documentation is added in dbf9e69d1f0b3293e36d97c6cf378bdfdb422fa8

LPTK replied:
Ok sure. Not sure it's worth it, but it's a detail.

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,104 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+      copy(decls = decls.view.mapValues(mt => mt.copy(parents = prt :: mt.parents)).toMap)
+    def propagate(top: Bool = true)(implicit raise: Raise): MethodDefs =
+      parents.map(_.propagate(false)).reduceOption(_.&(_)(tn, defns.keySet)).map(_.addParent(tn))
+        .foldRight(if (top) copy(parents = Nil, decls = Map.empty, defns = Map.empty) else copy(parents = Nil))(_ ++ _)
```
This `copy` call may be cleaner as:
```suggestion
        .foldRight(if (top) MethodDefs(tn, Nil, Map.empty, Map.empty) else copy(parents = Nil))(_ ++ _)
```

fo5for replied:
Addressed in dbf9e69d1f0b3293e36d97c6cf378bdfdb422fa8. (I wrote it this way to make the two branches more symmetric.)

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -20,60 +20,104 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+            S(mn -> MethodType(defns.head.level, N, tn :: Nil))
+        }
+      )
+    private def ++(that: MethodDefs): MethodDefs = {
```
Not sure why this `private def` is needed. It's only used once in `propagate` and does not seem to make a lot of sense on its own.

fo5for replied:
Addressed in dbf9e69d1f0b3293e36d97c6cf378bdfdb422fa8

LPTK commented:
(Sorry I keep on piling comments on this PR while you're trying to merge! 😛)

fo5for committed:
Require overriding defn for overriding method decl with defn

fo5for committed:
Cleanup and documentation

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +376,239 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+        def ss(mt: MethodType, bmt: MethodType)(implicit prov: TypeProvenance) =
+          constrain(subst(mt.toPT, targsMap).instantiate, subst(bmt.toPT, targsMap).rigidify)
+        def registerImplicit(nme: Str, mthTy: MethodType) = ctx.getMth(N, nme) match {
+          case S(MethodType(_, _, parents)) if parents.iterator.map(prt => Var(prt.name.decapitalize)).toSet
```
This could be written much as the more efficient `if { val bcs = ctx.allBaseClassesOf(tn.name); parents.forall(...) }`.

fo5for replied:
Addressed in c5f6b0817ec9905325a697a9281652782b2d2e30

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +376,239 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+          constrain(subst(mt.toPT, targsMap).instantiate, subst(bmt.toPT, targsMap).rigidify)
+        def registerImplicit(nme: Str, mthTy: MethodType) = ctx.getMth(N, nme) match {
+          case S(MethodType(_, _, parents)) if parents.iterator.map(prt => Var(prt.name.decapitalize)).toSet
+              .subsetOf(ctx.allBaseClassesOf(tn.name)) =>
```
This non-trivial condition really needs some doc. Be empathetic to the people who will be trying to understand your intention later on!! 😉
Despite the fact I _know_ the semantics of methods very well, I have no idea what this is checking...
```suggestion
              .subsetOf(ctx.allBaseClassesOf(tn.name)) =>
                // If [...] then we don't need to do anything
```

fo5for replied:
Addressed in c5f6b0817ec9905325a697a9281652782b2d2e30

LPTK replied:
Great, the comments make things infinitely clearer now, especially the example.

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +376,239 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+        val targsMap = td.targs.lazyZip(rigidtargs).toMap[SimpleType, SimpleType]
+        def ss(mt: MethodType, bmt: MethodType)(implicit prov: TypeProvenance) =
+          constrain(subst(mt.toPT, targsMap).instantiate, subst(bmt.toPT, targsMap).rigidify)
+        def registerImplicit(nme: Str, mthTy: MethodType) = ctx.getMth(N, nme) match {
```
`nme` is the name of what?

fo5for replied:
I have changed all uses of `nme` in `checkSubsume` to `mn` to clarify that it's the method name in c5f6b0817ec9905325a697a9281652782b2d2e30

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +376,239 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+              .subsetOf(ctx.allBaseClassesOf(tn.name)) =>
+          case S(MethodType(_, _, parents)) if parents.forall(prt =>
+              ctx.allBaseClassesOf(prt.name).contains(Var(tn.name.decapitalize))) =>
+            ctx.addMth(N, nme, mthTy)
```
```suggestion
            // If [...], then we need to register the new method
            ctx.addMth(N, nme, mthTy)
```

fo5for replied:
Addressed in c5f6b0817ec9905325a697a9281652782b2d2e30

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +376,239 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+            case (N, N) => N
+          }).foreach(ss(mt, _))
+          val mthTy = td.wrapMethod(mt)
+          if (!decls.isDefinedAt(nme)) {
```
```suggestion
          if (!decls.isDefinedAt(nme)) {
            // If the class does not declare that method explicitly (it is inherited),
            // we still want to make the method available
            // both for implicit method calls (of the form `x.Foo`)
            // and for explicit ones (`x.(TheClass.Foo)`)
```

fo5for replied:
Addressed in c5f6b0817ec9905325a697a9281652782b2d2e30. The examples for explicit vs. implicit calls are removed as they were documented elsewhere.

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -365,201 +377,260 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
+          constrain(subst(mt.toPT, targsMap).instantiate, subst(bmt.toPT, targsMap).rigidify)
+        def registerImplicit(mn: Str, mthTy: MethodType) = ctx.getMth(N, mn) match {
+          // If the currently registered method belongs to one of the base classes of this class,
+          // then we don't need to do anything
```
```suggestion
          // then we don't need to do anything.
          // This is because implicit method calls always default to the parent methods.
```

LPTK reviewed (approved):
Okay, great work! I think it's almost ready to merge now. I think it's best to rebase-merge it. But be sure to squash the last two commits ("Cleanup and documentation" and "More cleanup and documentation") into a single commit first.

fo5for commented:
I noticed that `codegen/Classes.mls` is failing after rebasing, with the following error:
```scala
class MyBox: Box[int] & { info: string }
  method Map f = Box (f this.inner) with { info = this.info }
  method Inc = MyBox { inner = this.inner + 1; info = this.info }
def MyBox inner info = MyBox { inner; info }
//│ ╔══[ERROR] Type mismatch in inherited method definition:
//│ ║  l.10: 	  method Map f = Box { inner = f this.inner }
//│ ║        	         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── expression of type `{inner: ?a}` does not match type `'a`
//│ ║  l.10: 	  method Map f = Box { inner = f this.inner }
//│ ║        	                     ^^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── but it flows into application with expected type `Box['a]`
//│ ║  l.10: 	  method Map f = Box { inner = f this.inner }
//│ ║        	                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── Note: constraint arises from type variable:
//│ ║  l.9: 	  method Map: (T -> 'a) -> Box['a]
//│ ║       	                               ^^
//│ ╟── from record type:
//│ ║  l.8: 	class Box[T]: { inner: T }
//│ ╙──     	              ^^^^^^^^^^^^
//| ...
```
It took me a while to understand what was going on. In the old version, methods are typed once, then substituted when derived classes inherit them. So after the shorthand `Box` constructor is defined, the body of `Box.Map` still refers to the original constructor. But now that methods are typed every time a derived class inherits them, the body of `Box.Map` refers to the shorthand constructor instead when the inherited methods of `MyBox` are typed...

Should I push it anyways so you can have a look at it?

LPTK commented:
Oh! Method _bodies_ should definitely not be retyped every time.

Now, I understand that this probably means it'd be simpler to just go back to the previous approach where things were typed once and then we used type substitution. But I wouldn't want you to spend much more time on this, so I think it's fine if we capture and store the `ctx` in each type definition, and retype method bodies based on that context instead. Does that seem feasible to you?

fo5for commented:
Yeah I was thinking of the same thing. Storing a bunch of `MutMap`s with mostly the same content seems like a waste of space, which can be avoided by returning a nested `Ctx` instead when `TypeDef`s are processed. But a long list would potentially need to be traversed when looking up the environment. Basically this would compromise either the space complexity of `TypeDef`s or the time complexity of lookups...

LPTK commented:
But the maximum depth of any inheritance hierarchy in real programs is very small, and does not grow linearly as the program grows at all. So traversing a list the size the of inheritance hierarchy depth upon lookup seems totally fine to me, and in fact it'll probably be faster than copying mutable maps around _or_ going full immutable.

On the other hand, storing plain context copies seems much easier so I'd encourage you to do that. Eventually we'll change it and do the right thing anyway. (If someone is reading this 20 years later and this still has not changed, and I'm  not dead, ping me.)

fo5for commented:
I got something kinda working, but it's very hacky, and ultimately storing a context isn't nice. I came up with another solution: how about we change `env` to map to a list (or vector) instead, and add an optional index value to `Var`, indicating the index it refers to in the list. Then we traverse the bodies of method definitions after registering the constructors to the context, and register the index of the last element in the list to all `Var`s. For `Var`s outside of method definitions, they would point to index `None`, and would just be typed to the last element in the list. This sounds relatively clean and doesn't require too much changes to the existing code.

LPTK commented:
I don't think I'd change the entire context and `Var` representations just to avoid doing something that should and will be done eventually.

If what you have now is too hacky, how much work would it be to just do it right? As in, only type methods once, and then do substitutions.

fo5for commented:
I started experimenting with changing `Var` and context, got too invested and completely missed your reply until I got it working. 😅 

Actually probably not too much, since the definitions are already stored in `mthEnv` for printing in `DiffTests`.

fo5for committed:
Fix retyping inherited method definitions

Inherited method definitions are still typed if they are defined in a
class after the current class.

fo5for commented:
Turns out doing it the right way is the easiest! 😅

fo5for commented:
Yikes I realized I forgot to do the substitution. No idea why it worked. Fixing now 😅

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -24,56 +24,97 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
-   *    For the nested context during method typing, the innermost values can also be `TraitTag`s.
+  /** Keys of `mthEnv`:
+   *  `L` represents the inferred types of method definitions. The first value is the parent name, and the second value is the method name.
+   *    The method definitions are *not* wrapped with the implicit `this` parameter, but is instead wrapped at `getMthDefn`.
```
```suggestion
   *    The method definitions are *not* wrapped with the implicit `this` parameter, but are instead wrapped at `getMthDefn`.
```

LPTK commented:
> No idea why it worked.

Whoops! The best thing to do would be to add a test that fails because of this, before fixing it.

LPTK commented on shared/src/main/scala/mlscript/Typer.scala:
```diff
@@ -491,8 +495,12 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
-                typeLetRhs(rec, nme.name, term)(thisCtx, raise, targsMap2),
+              term => ctx.getMthDefn(prt.name, nme.name, false)
+                .fold(typeLetRhs(rec, nme.name, term)(thisCtx, raise, targsMap2))(_.toPT match {
+                  case PolymorphicType(level, ProvType(underlying)) =>
```
This look pretty weird and ad-hoc. I'd like a comment explaining why it's necessary.

fo5for replied:
It's not *that* necessary, but it makes the error reporting nicer. I have added a comment to explain that.

LPTK reviewed (commented):
Note about the merging strategy: since you rebased on top of the main branch, your intermediate commits now fail because of the new `Classes` test. This is not ideal as intermediate commits should ideally all pass tests. 
Since your branch has meaningful commits over a long period, it would be fine to make a proper merge commit instead (rebase back to where you were, then make a merge commit after the fix). That, or we'll just squash the branch.

fo5for commented:
I think it's fine to squash the PR. There were some changes that I made and reverted afterwards. The weird class hierarchy of the two new classes also didn't make much sense.

> > No idea why it worked.
> 
> Whoops! The best thing to do would be to add a test that fails because of this, before fixing it.

Took me some time to find something that should fail but didn't!

LPTK committed:
Minor formatting
